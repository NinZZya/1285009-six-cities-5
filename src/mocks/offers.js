export default [
  {
    id: 0,
    city: {
      id: 1,
      name: `Paris`,
    },
    title: `Wood and stone place`,
    type: `Private room`,
    price: 228,
    rate: 0,
    bedroomsCount: 1,
    adultsCount: 1,
    features: [
      `Dishwasher`,
      `Baby seat`,
    ],
    host: {
      name: `Angelina`,
      avatar: `/img/avatar-angelina.jpg`,
      isPro: false,
    },
    images: [
      `/img/apartment-03.jpg`,
      `/img/apartment-02.jpg`,
    ],
    description: `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
    reviews: [
      {
        id: 0,
        user: {
          id: 3,
          name: `John`,
          avatar: `/img/avatar.svg`,
        },
        date: `2019-11-26`,
        rate: 2,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      },
      {
        id: 1,
        user: {
          id: 3,
          name: `John`,
          avatar: `/img/avatar.svg`
        },
        date: `2019-03-23`,
        rate: 4,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      }
    ],
    isPremium: false,
  },
  {
    id: 1,
    city: {
      id: 1,
      name: `Paris`,
    },
    title: `Canal View Prinsengracht`,
    type: `Apartment`,
    price: 97,
    rate: 3,
    bedroomsCount: 1,
    adultsCount: 3,
    features: [
      `Dishwasher`,
      `Towels`,
      `Fridge`,
      `Heating`,
      `Washing machine`,
      `Wi-Fi`,
      `Coffee machine`,
    ],
    host: {
      name: `John`,
      avatar: `/img/avatar.svg`,
      isPro: false,
    },
    images: [
      `/img/apartment-01.jpg`,
      `/img/apartment-02.jpg`,
    ],
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
    reviews: [
      {
        id: 2,
        user: {
          id: 4,
          name: `Ann`,
          avatar: `/img/avatar.svg`
        },
        date: `2018-09-22`,
        rate: 2,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      },
      {
        id: 3,
        user: {
          id: 1,
          name: `Max`,
          avatar: `/img/avatar-max.jpg`
        },
        date: `2018-10-24`,
        rate: 0,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      },
      {
        id: 4,
        user: {
          id: 4,
          name: `Ann`,
          avatar: `/img/avatar.svg`
        },
        date: `2019-08-16`,
        rate: 1,
        text: `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
      }
    ],
    isPremium: false,
  },
  {
    id: 2,
    city: {
      id: 2,
      name: `Paris`
    },
    title: `Beautiful & luxurious apartment at great location`,
    type: `Apartment`,
    price: 89,
    rate: 1,
    bedroomsCount: 2,
    adultsCount: 1,
    features: [],
    host: {
      name: `Angelina`,
      avatar: `/img/avatar-angelina.jpg`,
      isPro: false,
    },
    images: [
      `/img/apartment-01.jpg`
    ],
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
    reviews: [
      {
        id: 5,
        user: {
          id: 4,
          name: `Ann`,
          avatar: `/img/avatar.svg`,
        },
        date: `2019-06-07`,
        rate: 3,
        text: `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
      },
      {
        id: 6,
        user: {
          id: 1,
          name: `Max`,
          avatar: `/img/avatar-max.jpg`,
        },
        date: `2019-06-02`,
        rate: 1,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      }
    ],
    isPremium: false,
  },
  {
    id: 3,
    city: {
      id: 2,
      name: `Paris`,
    },
    title: `Wood and stone place`,
    type: `Private room`,
    price: 381,
    rate: 3,
    bedroomsCount: 1,
    adultsCount: 1,
    features: [
      `Fridge`,
      `Towels`,
    ],
    host: {
      name: `Angelina`,
      avatar: `/img/avatar-angelina.jpg`,
      isPro: false
    },
    images: [
      `/img/room.jpg`,
      `/img/apartment-02.jpg`,
      `/img/apartment-03.jpg`,
      `/img/studio-01.jpg`,
    ],
    description: `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
    reviews: [
      {
        id: 7,
        user: {
          id: 4,
          name: `Ann`,
          avatar: `/img/avatar.svg`,
        },
        date: `2018-11-08`,
        rate: 4,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      },
      {
        id: 8,
        user: {
          id: 2,
          name: `Angelina`,
          avatar: `/img/avatar-angelina.jpg`
        },
        date: `2019-02-23`,
        rate: 0,
        text: `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
      }
    ],
    isPremium: false,
  }
];
