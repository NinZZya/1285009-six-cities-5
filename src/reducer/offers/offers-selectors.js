import {createSelector} from 'reselect';
import NameSpace from '../name-space';
import {SortOffers} from '../../utils/utils';
import {getActiveCityId} from '../cities/cities-selectors';


const NAME_SPACE = NameSpace.OFFERS;


export const getOffersStatus = (state) => state[NAME_SPACE].status;
export const getOffersSortType = (state) => state[NAME_SPACE].sortType;
export const getOffers = (state) => Object.values(state[NAME_SPACE].offers);
export const getOffersMap = (state) => state[NAME_SPACE].offers;

export const getOfferStatus = (state) => state[NAME_SPACE].offerStatus;
export const getOffer = (state, id) => state[NAME_SPACE].offers[id];
export const getOfferReviewsStatus = (state) => state[NAME_SPACE].reviewsStatus;
export const getOfferReviews = (state) => state[NAME_SPACE].reviews;
export const getNearOffersStatus = (state) => state[NAME_SPACE].nearOffersStatus;
export const getNearOffers = (state) => state[NAME_SPACE].nearOffers;
export const getFavoritesOffersMap = (state) => state[NAME_SPACE].favorites;

export const getFavoritesOffers = (state) =>{
  const offers = Object.values(state[NAME_SPACE].favorites);
  return offers.reduce((favorites, offer) => {
    if (favorites[offer.city.name]) {
      favorites[offer.city.name].push(offer);
    } else {
      favorites[offer.city.name] = [offer];
    }

    return favorites;
  }, {});
};

export const getSortedCityOffers = createSelector(
    getOffers,
    getActiveCityId,
    getOffersSortType,
    (offers, cityId, sortType) => {
      const cityOffers = offers
        .filter((offer) => offer.city.id === cityId);
      return SortOffers[sortType](cityOffers);
    }
);

