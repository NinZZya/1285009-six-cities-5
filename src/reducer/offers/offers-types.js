export const CHANGE_OFFERS_STATUS = `CHANGE_OFFERS_STATUS`;
export const SET_OFFERS = `SET_OFFERS`;
export const CHANGE_OFFERS_SORT_TYPE = `CHANGE_OFFERS_SORT_TYPE`;
