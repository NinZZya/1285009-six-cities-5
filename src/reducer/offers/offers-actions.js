import * as OffersType from './offers-types';
import {makeActionCreator} from '../../utils/utils';


export const changeOffersStatus = makeActionCreator(OffersType.CHANGE_OFFERS_STATUS);
export const setOffers = makeActionCreator(OffersType.SET_OFFERS);
export const changeOffersSortType = makeActionCreator(OffersType.CHANGE_OFFERS_SORT_TYPE);
