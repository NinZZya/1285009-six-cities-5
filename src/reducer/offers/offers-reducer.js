import * as OffersType from './offers-types';
import {extend} from '../../utils/utils';
import {DataStatus, SortType} from '../../const';


const initialState = {
  status: DataStatus.LOADING,
  offers: {},
  favorites: {},
  sortType: SortType.POPULAR,
  offerStatus: DataStatus.LOADING,
  reviewsStatus: DataStatus.LOADING,
  reviews: [],
  nearOffersStatus: DataStatus.LOADING,
  nearOffers: [],
};


export default (state = initialState, action) => {
  switch (action.type) {
    case OffersType.CHANGE_OFFERS_STATUS:
      return extend(state, {
        status: action.payload,
      });
    case OffersType.SET_OFFERS:
      return extend(state, {
        offers: action.payload,
      });
    case OffersType.CHANGE_OFFERS_SORT_TYPE:
      return extend(state, {
        sortType: action.payload,
      });
    case OffersType.CHANGE_OFFER_STATUS:
      return extend(state, {
        offerStatus: action.payload,
      });
    case OffersType.CHANGE_OFFER_REVIEWS_STATUS:
      return extend(state, {
        reviewsStatus: action.payload,
      });
    case OffersType.SET_OFFER_REVIEWS:
      return extend(state, {
        reviews: action.payload,
      });
    case OffersType.CHANGE_NEAR_OFFERS_STATUS:
      return extend(state, {
        nearOffersStatus: action.payload,
      });
    case OffersType.SET_NEAR_OFFERS:
      return extend(state, {
        nearOffers: action.payload,
      });
    case OffersType.SET_FAVORITES_OFFERS:
      return extend(state, {
        favorites: action.payload,
      });
    default:
      return state;
  }
};
