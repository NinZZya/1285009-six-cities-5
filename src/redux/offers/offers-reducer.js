import * as OffersType from './offers-types';
import {LoadStatus, SortType, DEFAULT_CITY_ID} from '../../const';


const initialState = {
  status: LoadStatus.LOADING,
  offers: {},
  sortType: SortType.POPULAR,
  activeCityId: DEFAULT_CITY_ID,
  reviewStatus: LoadStatus.LOADING,
  reviews: [],
};

export default (state = initialState, action) => {
  switch (action.type) {
    case OffersType.CHANGE_OFFERS_STATUS:
      return Object.assign({}, state, {
        status: action.payload,
      });
    case OffersType.SET_OFFERS:
      return Object.assign({}, state, {
        offers: action.payload,
      });
    case OffersType.CHANGE_OFFERS_SORT_TYPE:
      return Object.assign({}, state, {
        sortType: action.payload,
      });
    case OffersType.CHANGE_ACTIVE_CITY_ID:
      return Object.assign({}, state, {
        activeCityId: action.payload,
      });
    case OffersType.CHANGE_OFFER_REVIEWS_STATUS:
      return Object.assign({}, state, {
        reviewStatus: action.payload,
      });
    case OffersType.SET_OFFER_REVIEWS:
      return Object.assign({}, state, {
        reviews: action.payload,
      });
    default:
      return state;
  }
};

