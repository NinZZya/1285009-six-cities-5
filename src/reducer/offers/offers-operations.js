import * as OffersAction from './offers-actions';
import * as CitiesAction from '../cities/cities-actions';
import {adaptOffersDataToClient} from '../../adapters/offers-adapter';
import {DataStatus} from '../../const';


export const loadOffersAsync = () => (dispatch, getState, api) => {
  return api.getOffers()
    .then((response) => {
      if (response.length) {
        const adaptData = adaptOffersDataToClient(response);
        const {offers, cities} = adaptData;
        dispatch(CitiesAction.updateCities(cities));
        dispatch(OffersAction.setOffers(offers));
        dispatch(OffersAction.changeOffersStatus(DataStatus.SUCCESS));
      } else {
        dispatch(OffersAction.setOffers({}));
        dispatch(OffersAction.changeOffersStatus(DataStatus.SUCCESS));
      }


    })
    .catch(() => {
      dispatch(OffersAction.changeOffersStatus(DataStatus.ERROR));
    });
};
