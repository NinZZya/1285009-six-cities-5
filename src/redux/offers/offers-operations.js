import * as OffersAction from './offers-actions';
import {LoadStatus} from '../../const';


export const loadOffersAsync = () => (dispatch, getOffers, api) => {
  return api.getOffers()
    .then((response) => {
      const offers = response.length ?
        api.adaptOffersToClient(response) :
        {};

      dispatch(OffersAction.setOffers(offers));
      dispatch(OffersAction.changeOffersStatus(LoadStatus.SUCCESS));
    });
};

export const loadReviewsAsync = (id) => (dispatch, getComments, api) => {
  dispatch(OffersAction.changeOfferReviewsStatus(LoadStatus.LOADING));
  return api.getReviews(id)
    .then((response) => {
      const reviews = response ? response : [];

      dispatch(OffersAction.setOfferReviews(reviews));
      dispatch(OffersAction.changeOfferReviewsStatus(LoadStatus.SUCCESS));
    });
};
