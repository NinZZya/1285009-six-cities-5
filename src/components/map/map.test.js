import React from 'react';
import renderer from 'react-test-renderer';
import Map from './map';
import offers from '@/mocks/offers';

const CENTER = offers[0];
const PINS = offers;
const ACTIVE_ID = 0;

describe(`Should Map render correctly`, () => {
  it(`Should Map render correctly with pins`, () => {
    const tree = renderer
      .create(
          <Map
            center={CENTER}
            pins={PINS}
            activeId={ACTIVE_ID}
          />,
          {createNodeMock: () => {
            return document.createElement(`div`);
          }}
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should Map render correctly without pins`, () => {
    const tree = renderer
      .create(
          <Map
            center={CENTER}
            pins={[]}
            activeId={ACTIVE_ID}
          />,
          {createNodeMock: () => {
            return document.createElement(`div`);
          }}
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
