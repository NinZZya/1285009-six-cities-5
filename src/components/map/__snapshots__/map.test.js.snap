// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Map render correctly Should Map render correctly with pins 1`] = `
<div
  id="map"
  style={
    Object {
      "height": "100%",
    }
  }
/>
`;

exports[`Should Map render correctly Should Map render correctly without pins 1`] = `
<div
  id="map"
  style={
    Object {
      "height": "100%",
    }
  }
/>
`;
